odoo/tools/misc.py

```diff
@@ -1821,6 +1821,44 @@ def verify_hash_signed(env, scope, payload):
     return None
 
 
+def binary_token(record, field_name, timestamp=None):
+    """Generate a token granting access for at least 14 days from the binary
+    routes (/web/content or /web/image) to the given record and field_name.
+    The token is deterministic within a 14-day period (even across different
+    days/months/years) to allow browser caching, and expires after maximum 28
+    days to prevent infinite access. Different record/field combinations will
+    expire at different times to prevent thundering herd problems."""
+    record.ensure_one()
+    if not timestamp:
+        now = datetime.datetime.now()
+        epoch = datetime.datetime(1970, 1, 1)
+        start_of_period = epoch + datetime.timedelta(days=((now - epoch).days // 14) * 14)
+        unique_str = f"{record._name}-{str(record.id)}-{field_name}"
+        days_offset = int(hashlib.sha256(f"d-{unique_str}".encode()).hexdigest(), 16)
+        seconds_offset = int(hashlib.sha256(f"s-{unique_str}".encode()).hexdigest(), 16)
```

Comment:
It's used to add more jitter over the whole day rather than always at a fixed time, it's not more or less necessary than the days offset.

For 1000+ users databases split into 14 days, I felt it still a lot of refresh at the exact same time if we don't also spread the time.